provider/oracle/environ.go

```diff
@@ -0,0 +1,731 @@
+// inside a oracle environ
+type EnvironAPI interface {
+	commonProvider.Instancer
+	commonProvider.InstanceAPI
+	commonProvider.Authenticater
+	commonProvider.Shaper
+	commonProvider.Imager
+	commonProvider.IpReservationAPI
+	commonProvider.IpAssociationAPI
+	commonProvider.IpNetworkExchanger
+	commonProvider.IpNetworker
+	commonProvider.VnicSetAPI
+
+	commonProvider.RulesAPI
+	commonProvider.AclAPI
+	commonProvider.SecIpAPI
+	commonProvider.IpAddressPrefixSetAPI
+	commonProvider.SecListAPI
+	commonProvider.ApplicationsAPI
+	commonProvider.SecRulesAPI
+	commonProvider.AssociationAPI
+
+	StorageAPI
+}
+
+// AvailabilityZones is defined in the common.ZonedEnviron interface
+func (o *oracleEnviron) AvailabilityZones() ([]common.AvailabilityZone, error) {
+	return []common.AvailabilityZone{
+		oraclenet.NewAvailabilityZone("default"),
+	}, nil
+}
+
+// InstanceAvailabilityzoneNames is defined in the common.ZonedEnviron interface
+func (o *oracleEnviron) InstanceAvailabilityZoneNames(ids []instance.Id) ([]string, error) {
+	instances, err := o.Instances(ids)
+	if err != nil && err != environs.ErrPartialInstances {
+		return nil, err
+	}
+	zones := make([]string, len(instances))
+	for idx, _ := range instances {
+		zones[idx] = "default"
+	}
+	return zones, nil
+}
+
+// newOracleEnviron returns a new oracleEnviron
+func newOracleEnviron(p *environProvider, args environs.OpenParams, client EnvironAPI) (env *oracleEnviron, err error) {
+	if client == nil {
+		return nil, errors.NotFoundf("oracle client")
+	}
+	if p == nil {
+		return nil, errors.NotFoundf("environ proivder")
+	}
+	env = &oracleEnviron{
+		p:      p,
+		spec:   args.Cloud,
+		cfg:    args.Config,
+		mutex:  &sync.Mutex{},
+		client: client,
+		clock:  clock.WallClock,
+	}
+	env.namespace, err = instance.NewNamespace(env.cfg.UUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	env.Firewaller = oraclenet.NewFirewall(env, client)
+	env.Networking = oraclenet.NewEnviron(client)
+
+	return env, nil
+}
+
+// PrepareForBootstrap is part of the Environ interface.
+func (o *oracleEnviron) PrepareForBootstrap(ctx environs.BootstrapContext) error {
+	if ctx.ShouldVerifyCredentials() {
+		logger.Infof("Logging into the oracle cloud infrastructure")
+		if err := o.client.Authenticate(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+// Bootstrap is part of the Environ interface.
+func (o *oracleEnviron) Bootstrap(ctx environs.BootstrapContext, args environs.BootstrapParams) (*environs.BootstrapResult, error) {
+	return common.Bootstrap(ctx, o, args)
+}
+
+// Create is part of the Environ interface.
+func (o *oracleEnviron) Create(params environs.CreateParams) error {
+	if err := o.client.Authenticate(); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// AdoptResources is part of the Environ interface.
+func (e *oracleEnviron) AdoptResources(controllerUUID string, fromVersion version.Number) error {
+	return nil
```

Comment:
Please leave a TODO here to update the juju-controller-uuid tags of resources. If this isn't done, destroy-controller and kill-controller won't do the right thing after model migration.